API improvements

hhromic opened the pull request:
Hi !
I improved some things of this great node package. In particular I made the following changes:
- Added a new `isPortOpen()` method for the MIDI input and output objects. This method returns `true` or false depending if the port is opened or not.
- Added a little bit more advanced logic for ports opening/closing. In particular `openPort()` and `closePort()` will throw an error if the port was already opened or closed.
- Moved the call to the `uv_async_init()` method from the `New()` method to the `openPort()` and `openVirtualPort()` methods. This makes the initialization of the event emitting thread to be delayed until you actually open the MIDI input port. Why? because if you don't do this the Node program will hang before exiting until you call the `closePort()`, method no matter if you opened the port at some point. An example of when you don't need to open the port is when scanning for available port names.
- Made the `closePort()` method of the MIDI input object to unregister the callback function from the underlying RtMidi input object. This makes no real effect on everyday usage of the API, but takes away a bug if you try to re-open any MIDI port using the same MIDI input object. There is still and underlying bug in RtMidi that makes re-using MIDI input objects impossible, however I'm trying to see how to fix it. This problem only affects MIDI input, not MIDI output objects. This probably because of the callback mechanism.

Cheers and hoping soon for a 0.9.3 version =).

drewish commented:
I'm not sure I like throwing an error when closing a closed port. Seems better to just return since the desired outcome has been achieved.

drewish commented:
Also it would be good to try to add some test coverage to the unit tests for these changes: https://github.com/justinlatimer/node-midi/blob/master/test/unit-tests.js

hhromic commented:
Hi, about the behaviour of closing the port while it's already closed, I was thinking the same too. The decision depends if you want to be able to detect if it was already closed or not. Many APIs in Unix uses a return value for this, and in JavaScript you normally would use a callback function. For me it's ok either way (throwing errors or going silently).

Perhaps we can make openPort() and closePort() return a boolean on success/fail? this way if someone cares on detecting state it is just a matter on checking this return value, otherwise it will go silently.

I will also add some test coverages for the changes. I'm also not sure how to proceed. should I cancel this PR, make these extra changes then issue a new PR?

hhromic commented:
Hi again,
I finally managed to fix some nasty bugs in RtMidi. I adopted a more graceful approach for detecting ports already opened when using ```openPort()```, ```openVirtualPort()``` and ```closePort()``` as suggested. Now these methods return ```true```/```false``` accordingly (see proposed changelog). I also added unit tests for the new method ```isOpenPort()``` and the new return value semantics.

PRPOSED CHANGELOG
* Added binding for the isPortOpeN() in RtMidi. This method returns true or false depending if the port is currently opened or not.
* The openPort(), openVirtualPort() and closePort() methods now return true or false depending if the port is already open. Also these methods will not try to re-open an already opened port.
* Moved UV thread initialization to the openPort() and openVirtualPort() methods in MIDI input objects. This makes the initialization of the event emitting thread to be delayed until you actually open the MIDI input port. Before this fix, your Node program will not finish if you don't issue a closePort(), i.e. when only using getPortName() and getPortCount(). Now it works properly.

KNOWN ISSUES
* MIDI input objects can't be re-opened twice (after closing). This is a bug in RtMidi yet to be investigated.

drewish commented on src/node-midi.cpp:
```diff
@@ -50,14 +51,13 @@ class NodeMidiOutput : public node::ObjectWrap
         if (!args.IsConstructCall()) {
-            return NanThrowTypeError("Use the new operator to create instances of this object.");
+            NanThrowTypeError("Use the new operator to create instances of this object.");
+            NanReturnUndefined();
```
If we throw do we need to return?

drewish commented on src/lib/RtMidi/RtMidi.cpp:
```diff
@@ -646,6 +646,12 @@ void MidiInCore :: openPort( unsigned int portNumber, const std::string portName
+    error( RtMidiError::WARNING, errorString_ );
+    return;
+  }
+
```
Did these changes get merged upstream?

drewish commented on test/unit-tests.js:
```diff
@@ -62,6 +62,14 @@ describe('midi.input', function() {
     });
+
+    it('should return true if not already open', function() {
+      input.openPort(0).should.be.Boolean.and.eql(true);
```
Is there a reason you're checking the type is Boolean? Any reason not to just check for `.eql(true)` or `.eql(false)`?

drewish commented:
I think it might be easier to chunk this up into a few separate PRs:
- We can bump the Nan version and look at those changes
- Upgrade rtmidi

And this PR can just focus on the changes to opening and closing ports.

drewish commented:
This would need a rebase since we've updated NAN in another PR.

justinlatimer commented:
Most of these changes have been cherry picked. The rest are out of date.

hhromic commented:
@justinlatimer thanks for that, I completely forgot about this PR after all this time. Many apologies.
I'm happy to see that this project is still being actively maintained. Cheers!
